File: labs/04/src/Shape/Ellipse.php
<?php

namespace App\Shape;

use App\Canvas\CanvasInterface;
use App\Color;
use App\Point;

class Ellipse extends Shape
{
    private Point $center;
    private float $hRadius;
    private float $vRadius;

    public function __construct(Point $center, float $hRadius, float $vRadius, string $color = Color::BLACK)
    {
        $this->center = $center;
        $this->hRadius = $hRadius;
        $this->vRadius = $vRadius;
        $this->setColor($color);
    }

    public function getCenter() : Point
    {
        return $this->center;
    }

    public function getHorizontalRadius() : float
    {
        return $this->hRadius;
    }

    public function getVerticalRadius() : float
    {
        return $this->vRadius;
    }

    public function draw(CanvasInterface $canvas) : void
    {
        $center = $this->getCenter();
        $width = $this->getHorizontalRadius();
        $height = $this->getVerticalRadius();
        $color = $this->getColor();
        $canvas->setColor($color);
        $canvas->drawEllipse($center, $width, $height);
    }
}

File: labs/09/lab9v2/lab9v2/src/shape/Ellipse.java
package shape;

import canvas.CanvasInterface;

import java.awt.*;

public class Ellipse extends Shape {
    public Ellipse(Frame frame, Color strokeColor, Color fillColor) {
        super(frame, strokeColor, fillColor);
    }

    @Override
    public void draw(CanvasInterface canvas) {
        Frame frame = this.getFrame();
        canvas.fillEllipse(frame.getLeft(), frame.getTop(), frame.getWidth(), frame.getHeight(), this.getFillStyle().getColor());
        canvas.setLineSize(3);
        canvas.setColor(this.getStrokeStyle().getColor());
        canvas.drawEllipse(frame.getLeft(), frame.getTop(), frame.getWidth(), frame.getHeight());
    }

    @Override
    public boolean contains(Point point) {
        Frame frame = this.getFrame();
        Point center = new Point(frame.getLeft() + frame.getWidth() / 2, frame.getTop() + frame.getHeight() / 2);
        double x2 = Math.pow(point.x - center.x, 2);
        double y2 = Math.pow(point.y - center.y, 2);
        double w2 = Math.pow(frame.getWidth() / 2, 2);
        double h2 = Math.pow(frame.getHeight() / 2, 2);

        return x2 / w2 + y2 / h2 <= 1;
    }
}
